### PR TITLE
Add replace move for optional clients

### LINKED ISSUE #882: `Replace` operator
#### Is your feature request related to a problem? Please describe
Our local search operators currently only act on (U, V) node pairs that are both present in the solution. We're missing an opportunity here to evaluate node pairs where V is absent using node operators that have an 'inplace' mechanism.

#### Describe the solution you'd like
A local search node operator named that can work with (U, V), where U is present and V is absent. One possibility that comes to mind is `SwapInPlace<N>`, which swaps `U, U+1, ..., U+N` with `V`.

I quickly prototyped the operator https://github.com/PyVRP/PyVRP/tree/swap-in-place, and the results show that this works pretty well:

Image

There are probably many edge cases that I haven't considered yet (e.g., groups).

### PR DESCRIPTION
**Pull request by N-Wouda**
This PR:

- [x] Closes #882.
- [x] Adds and passes relevant tests.
- [x] Has well-formatted code and documentation.

See also #884 for a similar node operator.

<details>

**Notes**:

Please read our [contributing guidelines](https://pyvrp.org/dev/contributing.html) first.
In particular:

- You must add tests when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.
- Finally, it is essential that all contributions in this PR are license-compatible with PyVRP's MIT license.
  Please check that this PR can be included into PyVRP under the MIT license.

</details>

### CODE REVIEW

**Comment by N-Wouda** [comment]
This definitely breaks something 😆. Let's figure that out after dinner.

**Comment by N-Wouda** [comment]
OK so this sometimes cycles (a cycle of improving moves). That should not happen since it implies infinite improvement, so I need to figure out what we're evaluating wrong.

File: pyvrp/cpp/search/LocalSearch.cpp [hunk 1]
```diff
@@ -277,8 +277,36 @@ void LocalSearch::applyOptionalClientMoves(Route::Node *U,
         update(route, route);
     }
 
-    if (!U->route())
+    if (uData.required && !U->route())
+    {
         insert(U, costEvaluator, uData.required);
+        return;
+    }
+
```
Comment (leonlan) [inline pyvrp/cpp/search/LocalSearch.cpp:288]: It's assumed that U is not in a route after this, but that's not the case. So the next loop will try to insert U's that are already in the solution.

Comment (leonlan) [reply pyvrp/cpp/search/LocalSearch.cpp:288]: In other words, we should shortcut before the loop if `U` is in a route.

Comment (N-Wouda) [reply pyvrp/cpp/search/LocalSearch.cpp:288]: Yeah, I had that locally already but that isn't it. There's something wrong in inplaceCost, I think

File: pyvrp/cpp/search/LocalSearch.cpp [hunk 2]
```diff
@@ -277,8 +277,36 @@ void LocalSearch::applyOptionalClientMoves(Route::Node *U,
...
+        if (!route)
+            continue;
+
+        if (insertCost(U, V, data, costEvaluator) < 0)
+        {
+            route->insert(V->idx() + 1, U);
+            update(route, route);
+            return;
+        }
+
```
Comment (leonlan) [inline pyvrp/cpp/search/LocalSearch.cpp:304]: If it has to do with `inplaceCost` I suspect it might have to do with swapping out required `V`s. We should only allow replacing if `V` is not required. 

```cpp
        ProblemData::Client const &clientData = data.location(vClient);
        if (clientData.required)  // then we cannot remove V.
            continue;
```

This can be a problem when there is a mix of optional and required clients, such as in the `okSmallPrizes` instance. Combined with the other shortcut, this passes all tests on my machine.

(I'm not sure yet _why_ this is needed though)

Comment (N-Wouda) [reply pyvrp/cpp/search/LocalSearch.cpp:304]: I think removing required clients might be the problem, yeah. We probably shouldn't do that here.

**Comment by leonlan** [comment]
I’m running a benchmark now. The performance gain seems to be a little less than the previous node operator version, but I prefer this approach because it’s much simpler. Will post the results in the afternoon after I get back home

File: pyvrp/cpp/search/LocalSearch.cpp (hunk 2 above)
Comment (N-Wouda) [reply pyvrp/cpp/search/LocalSearch.cpp:304]: That was it, thanks!

**Comment by N-Wouda** [comment]
Uhh sure but this isn't done yet 😆. The insert/inplace evals don't have to be exact, for example. We can probably template those with a `bool exact = true` param, like in the `CostEvaluator`.

**Comment by N-Wouda** [comment]
Allowing inexact evals with the primitives is way more work. I'm putting that in a stash for now.

**Comment by N-Wouda** [comment]
OK I think this is reasonably OK. I need to add a test for swapping but I have to go now, so that'll have to come later.

**Comment by leonlan** [comment]
image

Benchmark. PCVRPTW is solved consistently better with this swap move. I assume that with delta cost shortcutting, this can still shave off a few 0.0x% but I think it's already good as-is.

File: tests/search/test_LocalSearch.py [hunk 3]
```diff
@@ -441,6 +441,45 @@ def test_local_search_does_not_remove_required_clients():
     assert_(new_cost < sol_cost)
 
 
+def test_replacing_optional_client():
```
Comment (N-Wouda) [inline tests/search/test_LocalSearch.py:444]: Checked that this fails on current main
